internal/resources/providers/msgraph/provider.go

```diff
@@ -0,0 +1,91 @@
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package msgraph
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+	graph "github.com/microsoftgraph/msgraph-sdk-go"
+	graphcore "github.com/microsoftgraph/msgraph-sdk-go-core"
+	"github.com/microsoftgraph/msgraph-sdk-go/models"
+	"github.com/microsoftgraph/msgraph-sdk-go/serviceprincipals"
+
+	"github.com/elastic/cloudbeat/internal/resources/providers/azurelib/auth"
+)
+
+type ProviderAPI interface {
+	ListServicePrincipals(context.Context) ([]*models.ServicePrincipal, error)
+}
+
+type provider struct {
+	log    *logp.Logger
+	client interface {
+		ServicePrincipals() *serviceprincipals.ServicePrincipalsRequestBuilder
+	}
+}
+
+// Second argument is scopes. Leave nil, then it selects default; Adjust if in trouble
+// Docs: https://learn.microsoft.com/en-us/graph/sdks/create-client?from=snippets&tabs=go
+func NewProvider(log *logp.Logger, azureConfig auth.AzureFactoryConfig) (ProviderAPI, error) {
+	// Requires 'Directory.Read.All' API permission
+	c, err := graph.NewGraphServiceClientWithCredentials(azureConfig.Credentials, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error creating MS Graph client: %w", err)
+	}
+
+	p := &provider{
+		log:    log.Named("msgraph"),
+		client: c,
+	}
+
+	return p, nil
+}
+
+// Docs:
+// - https://github.com/microsoftgraph/msgraph-sdk-go
+// - https://learn.microsoft.com/en-us/graph/api/serviceprincipal-list?view=graph-rest-beta&tabs=go
+// - https://learn.microsoft.com/en-us/graph/sdks/paging?tabs=go
+func (p *provider) ListServicePrincipals(ctx context.Context) ([]*models.ServicePrincipal, error) {
+	requestConfig := &serviceprincipals.ServicePrincipalsRequestBuilderGetRequestConfiguration{}
+
+	response, err := p.client.ServicePrincipals().Get(ctx, requestConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error listing Azure Service Principals: %w", err)
+	}
+
+	pageIterator, err := graphcore.NewPageIterator[*models.ServicePrincipal](
+		response,
+		p.client.ServicePrincipals().RequestAdapter,
+		models.CreateServicePrincipalCollectionResponseFromDiscriminatorValue,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error paging Azure Service Principals: %w", err)
+	}
+
+	items := []*models.ServicePrincipal{}
+	err = pageIterator.Iterate(ctx, func(pageItem *models.ServicePrincipal) bool {
+		items = append(items, pageItem)
+		return true // to continue the iteration
+	})
+	if err != nil {
+		p.log.Errorf("error iterating over Service Principals: %v", err)
```

Comment:
shouldn't we return the error if principals are not returned?

Reply:
This is a log, because we might still get some resources. The pagination failed, but `items` slice has been populated. We try to get the best scenario for us - getting data, even if incomplete.